### PR TITLE
🐦 Migratory Bird Treaty Act: Refactor the Redux store with a proposed migration mechanism

### PR DESCRIPTION
**Pull request by mhluongo**
> [The **Migratory Bird Treaty Act of 1918** (MBTA), codified at 16 U.S.C. §§ 703–712 (although §709 is omitted), is a United States federal law, first enacted in 1918 to implement the convention for the protection of migratory birds between the United States and Canada .[1] The statute makes it unlawful without a waiver to pursue, hunt, take, capture, kill, or sell nearly 1,100 species of birds listed therein as migratory birds. The statute does not discriminate between live or dead birds and also grants full protection to any bird parts including feathers, eggs, and nests. A March 2020 update of the list increased the number of species to 1,093.](https://en.wikipedia.org/wiki/Migratory_Bird_Treaty_Act_of_1918) — Wikipedia

We need to do some serious refactoring to the Redux store, but we haven't settled on a migration mechanism! After a conversation I had with @greg-nagy today, I decided to push the thing I've had in my head.

Here, I've made a small refactor to the Redux store as well as `PreferenceService` necessary for multi-network support. I've included migrations that I believe are production-ready, and tested them in a couple extension upgrade paths.

Interested to hear whether I'm headed in the right direction. I tried to stick with a small but central store change to help keep discussion focused.

### CODE REVIEW

**Review by greg-nagy — APPROVED** [review]
On migration I see this error message, but everything seems to be working fine

CleanShot 2021-12-28 at 15 19 59


This is a mind boggling cross cutting PR! Really impressed on all the aspects that have been handled!

File: background/redux-slices/selectors/uiSelectors.ts [hunk 1]
```diff
@@ -33,16 +33,17 @@ export const selectShowingActivityDetail = createSelector(
...
 
 export const selectCurrentAccount = createSelector(
-  (state: RootState) => state.ui.currentAccount,
-  ({ addressNetwork: { address }, truncatedAddress }) => ({
+  (state: RootState) => state.ui.selectedAccount,
+  ({ address, network }) => ({
     address,
-    truncatedAddress,
+    network,
+    truncatedAddress: address.toLowerCase().slice(0, 7),
```
[the remark this answers is not in the document]
Comment (mhluongo) [reply background/redux-slices/selectors/uiSelectors.ts:40]: I've been pushing for it myself, but we have like 3 truncate utils and tbh need a cross-cutting formatting PR... so I kicked the can :grimacing:

**Comment by mhluongo** [comment]
Hmm, that error message isn't *bad* but it does mean I missed deleting a key in that migration :thinking: 

I'll follow up with a new migration to handle it.

**Comment by greg-nagy** [comment]
I don't think that's the case. I am inclined to think that the reducer code might not have been updated by the time it gets the new state. 

This might be a race condition or caching issue. 

`selectedAccount` was created in the migration (it's the new key) and the error says it was found in the previous state.

**Pull request merged**